### PR TITLE
New root-scoped `RootFeatureRegistryService`.

### PR DESCRIPTION
**Pull request by davidfestal**
## Hey, I just made a Pull Request!

This PR introduces a new root-scoped `RootFeatureRegistryService` that allows getting the list of all the registered features (plugin or module) in the backend application.

A default service implementation is provided  (which simply gathers the list of all features added in the local backstage application). But it can be overridden by another implementation that would use some other mechanism to support distributed backends (as the discovery service already needs to be overridden in such cases).

This PR tries to implement task 1 of RFC [17538](https://github.com/backstage/backstage/issues/17538

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * The RootFeatureRegistryServiceImplementation is an abstract class that a RootFeatureRegistryService
+ * would extend, so that it can be fed with the list of registered features during
+ * application initialization.
+ *
+ * @public
+ */
+export abstract class RootFeatureRegistryServiceImplementation {
+  abstract setFeatures(features: BackendFeatureRegistration[]): void;
```
Comment (sennyeya) [inline packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts:40]: Thoughts on a `register`/`unregister` interface? `unregister` being unused in the default implementation, but possibly relevant to the distributed version. 

`unregister` would make me think we'd want this to be an orchestration layer as well, which seems odd. Maybe just `register` for now and we can add an optional `unregister` in the future when the distributed work starts.

**Review by sennyeya — COMMENTED** [review]
Thanks for this! Some small comments, I also reached out through Discord DMs with additional context about the distributed approach.

File: packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * The RootFeatureRegistryService is a root-scoped service used to provide a mechanism for backend
+ * plugins to discover the list of backend features registered in the application.
+ *
+ * Implementations of the registry API can be as simple as gathering the list of registered features
+ * at application initialization, but could also query a separate service.
+ *
+ * @public
+ */
+export interface RootFeatureRegistryService {
+  getFeatures(): BackendFeatureRegistration[];
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts]: Fixed

File: packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts (hunk 1 above)
Comment (sennyeya) [reply packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts:40]: We chatted offline, let's keep the current `setFeatures`.

Comment (davidfestal) [reply packages/backend-plugin-api/src/services/definitions/RootFeatureRegistryService.ts:40]: Not sure this would really be useful for in this context.
This method is called in the `BackendInitializer` after iterating through all the features and initializing all the InternalXXXFeatureRegistrations in the #doStart() method.

**Review by sennyeya — APPROVED** [review]

**Comment by davidfestal** [comment]
@freben wdyt of this PR ? It is still missing an additional approval to get merged.

**Review by Rugvip — CHANGES_REQUESTED** [review]
This is a tricky one, because it's the right direction but we don't know if we want to have this incremental step exist since it doesn't work well with scaled setups with multiple deployments. To clarify a bit, I'm talking about this kind of deployment architecture: https://backstage.io/docs/backend-system/building-backends/index#split-backend-deployments-architecture-example.

If we take that exact graph as an example, plugins deployed in backend A would only see `app`, `auth` and `proxy` in this service, but what they actually want is the full list of plugins. This kind of service to enumerate the features installed in the same Backstage backend instance is of course part of the solution, but it is **not** what we want plugins to consume in the end.

What we need in the end is a way for plugins to discovery all other available plugins, for example it might an extension of the `DiscoveryService`:

```ts
export interface DiscoveryService {
  listPlugins(): Promise<{ plugins: Array<{id: string}>}>
  // ...
}
```

That's the interface that we'd want plugins to use in practice, rather than one geared toward introspecting the local instance.

---

We actually happened to discuss this exact thing last week. Conclusion then was that it would be best to ship a full discovery solution at once, rather than shipping the local discovery first. Otherwise we risk having plugins adopt the partial solution and either remain broken for the scaled deployment scenario, or need to migrate in the future.

TL;DR: We'd like to take this all the way to extending the discovery API for plugins to use at once, however trivial that implementation might be for now.

**Comment by davidfestal** [comment]
> This is a tricky one, because it's the right direction but we don't know if we want to have this incremental step exist since it doesn't work well with scaled setups with multiple deployments. To clarify a bit, I'm talking about this kind of deployment architecture: https://backstage.io/docs/backend-system/building-backends/index#split-backend-deployments-architecture-example.
> 
> If we take that exact graph as an example, plugins deployed in backend A would only see `app`, `auth` and `proxy` in this service, but what they actually want is the full list of plugins. This kind of service to enumerate the features installed in the same Backstage backend instance is of course part of the solution, but it is **not** what we want plugins to consume in the end.
> 
> What we need in the end is a way for plugins to discovery all other available plugins, for example it might an extension of the `DiscoveryService`:
> 
> ```ts
> export interface DiscoveryService {
>   listPlugins(): Promise<{ plugins: Array<{id: string}>}>
>   // ...
> }
> ```
> 
> That's the interface that we'd want plugins to use in practice, rather than one geared toward introspecting the local instance.
> 
> We actually happened to discuss this exact thing last week. Conclusion then was that it would be best to ship a full discovery solution at once, rather than shipping the local discovery first. Otherwise we risk having plugins adopt the partial solution and either remain broken for the scaled deployment scenario, or need to migrate in the future.
> 
> TL;DR: We'd like to take this all the way to extending the discovery API for plugins to use at once, however trivial that implementation might be for now.

I might be missing something here, but it seems to me that, with my PR, nothing prevents anyone to write an implemetation of the `FeatureRegistry` service that would return the whole list of plugins / modules in a split deployments architecture.
That's the whole point of using a service that has a default implementation (in case of simple deployment) but can be overridden.
Such an overridden service would receive the list of features being installed locally, but this of course doesn't prevent it to grab any other information from outside to complete this list and return the whole list of features provided by the split-deployment application.

Let me quote something mentioned in the [Split backend docs](https://backstage.io/docs/backend-system/building-backends/index/#split-into-multiple-backends):

> You'll need to manually configure the two backends with custom implementations of the DiscoveryService and have them return the correct URLs for each other. Likewise, you'll also need to provide a custom implementation of the DiscoveryApi in the frontend, unless you surface the two backends via a proxy that handles the routing instead.

Obviously with the proposed PR, you would also need to provide a custom implementation of the `FeatureRegistry` service  that would return the whole list of features spanning the various deployments.

I don't see  how the minimalist service proposed here prevents building such an implementation on top of it.

**Comment by davidfestal** [comment]
Or should I create a distinct PR that would just add the `listPlugins` in the DiscoverService, and would, in the default implementation, feed it with the list of plugins ID in the BackendInitializer (the same it is done in this PR) ?

**Comment by davidfestal** [comment]
> Or should I create a distinct PR that would just add the `listPlugins` in the DiscoverService, and would, in the default implementation, feed it with the list of plugins ID in the BackendInitializer (the same it is done in this PR) ?

BTW is there a reason why the `DiscoverService` is not a root service ?

**Comment by davidfestal** [comment]
@Rugvip Please could you provide your thoughts about my last comments, so that I would know which direction to take in order to have this feature (listing backend plugins and possibly modules) available upstream ? We need it in the context of the migration of the Janus Showcase to the new backend system.

**Comment by davidfestal** [comment]
> Or should I create a distinct PR that would just add the `listPlugins` in the DiscoverService, and would, in the default implementation, feed it with the list of plugins ID in the BackendInitializer (the same it is done in this PR) ?

I have another proposal: do not add a new service for now, and simply allow any root service to receive the list of local features by simply implementing an interface (`BackendFeatureRegistrationObserver`). The only change it requires is that the `BackendInitializer` would provide the list of features to all such observing services.

This is a minimal and non-invasive implementation which unblock all users whose need is only to get the list of local features, and leaves for later the full implementation of an enhanced Discovery service.

I created an alternate PR for this: https://github.com/backstage/backstage/pull/22637  and would appreciate review and merge if possible.

**Comment by davidfestal** [comment]
New PR about supportting discovery in Split Deployments (https://github.com/backstage/backstage/pull/22671) is based on this one.
Could then this PR be merged already as a first step towards the more complete solution proposed in PR https://github.com/backstage/backstage/pull/22671 ?

**Comment by davidfestal** [comment]
@Rugvip in [your review comment](https://github.com/backstage/backstage/pull/22332#pullrequestreview-1838735208) you write :

> What we need in the end is a way for plugins to discovery all other available plugins, for example it might an extension of the DiscoveryService:
> 
> export interface DiscoveryService {
>   listPlugins(): Promise<{ plugins: Array<{id: string}>}>
>   // ...
> }
> That's the interface that we'd want plugins to use in practice, rather than one geared toward introspecting the local instance.
> 

But it seems to me, as the [Split Deployments PR](https://github.com/backstage/backstage/pull/22671) mentioned [in the comment above](https://github.com/backstage/backstage/pull/22332#issuecomment-1926921446) illustrates, that the `FeatureRegistryService` I propose here, which contains **local** feature registrations, is not at all incompatible with adding a `listPlugins()` method to the `DiscoveryService` to list all the plugin ids (local and remote).
In fact they seem to by quite distinct in purpose and complementary.
That's why I don't understand why you write:

> This is a tricky one, because it's the right direction but we don't know if we want to have this incremental step exist since it doesn't work well with scaled setups with multiple deployments 

Do you think that renaming the `RootFeatureRegistryService` to `RootLocalFeatureRegistryService` would make it clearer ?

**Review by Rugvip — CHANGES_REQUESTED** [review]
See https://github.com/backstage/backstage/pull/22671#pullrequestreview-1877209643 for updates on the other PR. The state is that we need a bit more upfront design work to agree on how to weave this all together, since it's a very critical part of the infra that can fail pretty catastrophically.

We also still want to avoid shipping a partial discovery solution as mentioned in https://github.com/backstage/backstage/pull/22332#pullrequestreview-1838735208. Tl;dr is that we don't want to introduce any APIs that plugins aren't supposed to be using long-term. It's alright if there are gaps in the implementation, but there can't be gaps in the design.
